### ISSUE TITLE
Add method to get a cached Population

### ISSUE THREAD

**Issue by adrien-berchet** [BlueBrain/NeuroM#928]
After using the new version of Population in several places, I realize that it would be nice to have a new method to ensure we are working on a cached Population. Because when I have a function that has to call `iter_neurites` or `iter_sections` several times and then use topological functions, I currently have to do this to ensure morphologies are kept in memory (because I don't know if the user will pass a cached population or not):

```python
def function_that_needs_a_cached_pop(pop):
    pop_ensure_cache = Population([i for i in pop.neurons])
    # ... do stuff on the pop_ensure_cache object
```

A method like `pop.get_cached_pop()` would be better I think.
I think it would be quite easy to implement, probably like this:

```python
class Population:
    # ...
    def __init__(...):
        # ...
        self._is_cached = cache
        # ...

    # ...
    def get_cached_pop():
        if self._is_cached:
            return self
        else:
            return Population(self.neurons)
    # ...
```

The previous function could thus become:

```python
def function_that_needs_a_cached_pop(pop):
    pop_ensure_cache = pop.get_cached_pop()
    # ... do stuff on the pop_ensure_cache object
```

WDYT @arnaudon @asanin-epfl ?

**Comment by adrien-berchet** [comment]
For example we want to do something like this in TNS (though it is done in a more complicated way):
```python
    from collections import defaultdict
    from neurom.core.neuron import iter_neurites
    neurites_by_type = defaultdict(list)

    for neurite in iter_neurites(NEU):
        neurites_by_type[neurite.type].append(neurite)

    for neurite_type in neurites_by_type:
        for neurite in neurites_by_type[neurite_type]:
            for section in neurite.iter_sections():
                print(section)  # actually do stuff on the section depending on the value of neurite_type
```
I guess it is done like this because we don't know which types can be given.
But anyway, this kind of processing needs a cached population.

**Comment by asanin-epfl** [comment]
Why have a method for making population cached if you can create it cached?

About your example. As we saw recently in #927 such calculations are slow and memory consuming. I would recommend to avoid such pattern in general. "do stuff on the section" can be done during iterating over sections. No need to store them all.

For now I can offer that Population is created as cached by default. `morph_stats` will be using a non-cached population internally.

**Comment by adrien-berchet** [comment]
After some more thinking, it is indeed better to avoid this kind of things so I updated TNS to iterate on the neurites only once.
In the end I think we can keep the Pouplation as it is since if a function does not work with a not cached Population it probably means that it is not correctly implemented, so the exception can act as a big warning for that.

**Comment by asanin-epfl** [comment]
ok, I'm closing the issue then. Still if you want, Population can be made cached by default.
